python/arcticdb/adapters/storage_fixture_api.py

```diff
@@ -0,0 +1,155 @@
+    out = ArcticMemoryConfig(cfg, Defaults.ENV)[name]
+    used[name] = out
+    return out
+
+
+class ArcticUriFields(Enum):
+    """For use with ``replace_uri_field()`` below"""
+    HOST = "HOST"
+    USER = "USER"
+    PASSWORD = "PASSWORD"
+    BUCKET = "BUCKET"
+
+    def __str__(self):
+        return self.value
+
+
+class StorageFixture(AbstractContextManager):
+    """Manages the life-cycle of a piece of storage and provides test facing methods:"""
+
+    _FIELD_REGEX: Dict[ArcticUriFields, Pattern]
+    """Used by ``replace_uri_field()``."""
+
+    arctic_uri: str
+    """The URI of this Storage for use with the ``Arctic`` constructor"""
+
+    generated_libs: Dict[str, NativeVersionStore] = {}
+
+    def __enter__(self):
+        """Fixtures are typically set up in __init__, so this just returns self"""
+        return self
+
+    @abstractmethod
+    def create_test_cfg(self, lib_name: str) -> EnvironmentConfigsMap:
+        """Creates a new storage config instance.
+        If ``lib_name`` is the same as a previous call on this instance, then that storage should be reused."""
+
+    def create_version_store_factory(self, default_lib_name: str):
+        """Returns a function that takes optional library options and produces ``NativeVersionStore``s"""
+        return functools.partial(
+            _version_store_factory_impl, self.generated_libs, self.create_test_cfg, default_lib_name
+        )
+
+    @abstractmethod
+    def set_permission(self, *, read: bool, write: bool):
+        """Makes the connection to the storage have the given permissions. If unsupported, call ``pytest.skip()``.
+        See ``set_enforcing_permissions`` below."""
+
+    @classmethod
+    def replace_uri_field(cls, uri: str, field: ArcticUriFields, replacement: str):
+        regex = cls._FIELD_REGEX[field]
+        match = regex.search(uri)
+        assert match, f"{uri} does not have {field}"
+        return f"{uri[:match.start(2)]}{replacement}{uri[match.end(2):]}"
+
+
+class StorageFixtureFactory(AbstractContextManager):
+    """For ``StorageFixture``s backed by shared/expensive resources, implement this class to manage those."""
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Properly clean up the fixture. This should be safe to be called multiple times."""
+
+    @property
+    def enforcing_permissions(self):
+        """Indicates whether the Factory will create Fixtures that enforces permissions.
+
+        If supported (settable), the new value should affect subsequent calls to ``create_fixture()``
+        and possibly existing fixtures."""
+        return False
+
+    @contextmanager
+    def enforcing_permissions_context(self, set_to=True):
+        saved = self.enforcing_permissions
+        try:
+            self.enforcing_permissions = set_to
+            yield
+        finally:
+            self.enforcing_permissions = saved
+
+    @abstractmethod
+    def create_fixture(self) -> StorageFixture:
+        ...
```

Comment:
`pass` not `...` consistent with rest of codebase

Reply:
It's [customary](https://docs.python.org/3/library/abc.html) to use `...` in `@abstractmethod` as it indicates something needs to be filled in.